### PR TITLE
[TS] LPP-46400 > LPS-162282 User enumeration attack with password expiration

### PR DESCRIPTION
**Pull request by chileces**
Dear Ms. To Trinh,
Our customer reported an issue related to application security "User enumeration attack with password expiration". This PR is to fix that issue, please help to review this and leave your comment if any.
Solution:
Check PasswordPolicy after Post-authentication pipeline Success.

Thank you very much,
Chi le
Previous PR:
https://github.com/brianchandotcom/liferay-portal/pull/122940

### CODE REVIEW

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 1]
```diff
@@ -91,6 +98,69 @@ public class UserLocalServiceTest {
...
+				null, null);
+		}
+		catch (PortalException portalException) {
+			user = _userLocalService.getUser(user.getUserId());
+
+			Assert.assertEquals(
+				failedLoginAttempts + 1, user.getFailedLoginAttempts());
+		}
+	}
+
```
Comment (tototrinh) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:163]: I think we could turn both into 1 test method. What do you think?

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 2]
```diff
@@ -91,6 +98,69 @@ public class UserLocalServiceTest {
...
+			_userLocalService.authenticateByEmailAddress(
+				user.getCompanyId(), user.getEmailAddress(), password, null,
+				null, null);
+		}
+		catch (PortalException portalException) {
+			user = _userLocalService.getUser(user.getUserId());
+
+			Assert.assertEquals(
+				failedLoginAttempts + 1, user.getFailedLoginAttempts());
+		}
```
Comment (tototrinh) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:161]: I think no need to call _userLocalService.authenticateByEmailAddress() again. Just using 
```
_userLocalService.fetchUser(user.getUserId());
Assert.assertEquals(1, newUser.getFailedLoginAttempts());
```

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 3]
```diff
@@ -91,6 +98,69 @@ public class UserLocalServiceTest {
...
+		_userLocalService.authenticateByEmailAddress(
+			user.getCompanyId(), user.getEmailAddress(), password, null, null,
+			null);
+	}
+
+	@Test
+	public void testAuthenticateByEmailAddressWrongPasswordWithPasswordPolicy()
+		throws Exception {
+
+		String newPassword = RandomTestUtil.randomString();
```
Comment (tototrinh) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:130]: No need to declare a variable. Can be used inline.

**Review by tototrinh — COMMENTED** [review]
@chileces, I have left some comments. Please take a look and feel free to ask me if you have any questions. Thanks.

**Comment by chileces** [comment]
i forced push some changes base on your comment, please help to review it

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 4]
```diff
@@ -91,6 +97,44 @@ public class UserLocalServiceTest {
...
+
+		try {
+			_userLocalService.authenticateByEmailAddress(
+				user.getCompanyId(), user.getEmailAddress(), newPassword, null,
+				null, null);
+		}
+		catch (PasswordExpiredException passwordExpiredException) {
+			Assert.assertEquals(
+				failedLoginAttempts + 1, user.getFailedLoginAttempts());
+		}
```
Comment (tototrinh) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:135]: I think it should be:
```
	catch (PortalException portalException) {
		Assert.assertEquals(
			PasswordExpiredException.class, portalException.getClass());
	}
```
and should move 
```
	Assert.assertEquals(
		failedLoginAttempts + 1, user.getFailedLoginAttempts());
```
out of the catch block and should fetch the user again before Assert.assertEquals.

**Review by tototrinh — COMMENTED** [review]
Hi @chileces, please see my comments. Thanks.

**Review by tototrinh — COMMENTED** [review]
Please take look at these comments. Thanks.

**Comment by ces-totrinh** [comment]
LGTM. Thanks.
Forwarded here: https://github.com/liferay-appsec/liferay-portal/pull/849

**Pull request closed without merging**